CGAL_Core/include/CGAL/CORE/BigInt.h

```diff
@@ -409,73 +459,99 @@ inline int cmpabs(const BigInt& a, const BigInt& b) {
 inline long longValue(const BigInt& a) {
   return a.longValue();
 }
+
 /// ulongValue
 inline unsigned long ulongValue(const BigInt& a) {
+    assert(a >= BigInt(0));
   return a.ulongValue();
 }
+
 /// doubleValue
 inline double doubleValue(const BigInt& a) {
   return a.doubleValue();
 }
 //@}
 
+/*
+
 /// \name File I/O Functions
 //@{
 /// read from file
+
 void readFromFile(BigInt& z, std::istream& in, long maxLength = 0);
 /// write to file
 void writeToFile(const BigInt& z, std::ostream& in, int base=10, int charsPerLine=80);
 //@}
+*/
+
 
 /// \name Misc Functions
 //@{
 /// isEven
 inline bool isEven(const BigInt& z) {
-  return mpz_even_p(z.get_mp());
+  return bit_test(z.get_mp(),0) == 0;
 }
 /// isOdd
 inline bool isOdd(const BigInt& z) {
-  return mpz_odd_p(z.get_mp());
+  return bit_test(z.get_mp(),0) == 1;
 }
 
 /// get exponent of power 2
 inline unsigned long getBinExpo(const BigInt& z) {
-  return mpz_scan1(z.get_mp(), 0);
+    if (z.get_mp().is_zero()) {
+        return (std::numeric_limits<unsigned long>::max)();
+    }
+    return lsb(abs(z.get_mp()));
 }
+
 /// get exponent of power k
-inline void getKaryExpo(const BigInt& z, BigInt& m, int& e, unsigned long k) {
-  mpz_t f;
-  mpz_init_set_ui(f, k);
-  m.makeCopy();
-  e = mpz_remove(m.get_mp(), z.get_mp(), f);
-  mpz_clear(f);
+inline void getKaryExpo(const BigInt& z, BigInt& m, int& e, unsigned long uk) {
+    BigInt k(uk), q, r;
+    e = 0;
+    m = z;
+    m.makeCopy();
+    for(;;) {
+        divide_qr(m.get_mp(), k.get_mp(), q.get_mp(), r.get_mp());
+        if (!r.get_mp().is_zero()) break;
+        m.get_mp() = q.get_mp();
+        ++e;
+    }
 }
 
 /// divisible(x,y) = "x | y"
 inline bool isDivisible(const BigInt& x, const BigInt& y) {
-  return mpz_divisible_p(x.get_mp(), y.get_mp()) != 0;
+    BigInt z, r;
+    divide_qr(x.get_mp(), y.get_mp(), z.get_mp(), r.get_mp());
```

Comment:
integer_modulus?

Reply:
Ah, indeed, `std::enable_if<boost::multiprecision::detail::is_integral<Integer>::value`. I think it would be worth opening an issue so they clarify their doc. The only hint there is the fact that the argument is passed by value...